File: python/tutorials/tour-of-orchestration-python/app/timers/service.py
import restate
from datetime import timedelta

from app.utils import init_payment, cancel_payment
from app.types import PaymentRequest, PaymentResult, ConfirmationRequest


payments_with_timeout = restate.Service("PaymentsWithTimeout")


@payments_with_timeout.handler()
async def process(ctx: restate.Context, req: PaymentRequest) -> PaymentResult:
    confirmation_id, confirmation_promise = ctx.awakeable(type_hint=PaymentResult)

    payment_id = str(ctx.uuid())
    pay_ref = await ctx.run_typed(
        "pay",
        init_payment,
        req=req,
        payment_id=payment_id,
        confirmation_id=confirmation_id,
    )

    # Race between payment confirmation and timeout
    match await restate.select(
        confirmation=confirmation_promise, timeout=ctx.sleep(timedelta(seconds=30))
    ):
        case ["confirmation", result]:
            return result
        case _:
            # Cancel the payment with external provider
            await ctx.run_typed("cancel-payment", cancel_payment, pay_ref=pay_ref)
            return PaymentResult(
                success=False, transaction_id=None, error_message="Payment timeout"
            )


@payments_with_timeout.handler()
async def confirm(ctx: restate.Context, confirmation: ConfirmationRequest) -> None:
    ctx.resolve_awakeable(confirmation.id, confirmation.result)

File: rust/tutorials/tour-of-orchestration-rust/src/timers/service.rs
use restate_sdk::{Context, Service, Result, RestateError, Awakeable};
use serde::{Deserialize, Serialize};
use std::time::Duration;
use uuid::Uuid;
use tokio::time::timeout;

#[derive(Deserialize)]
struct PaymentRequest {
    amount: i32,
    currency: String,
    customer_id: String,
    order_id: String,
}

#[derive(Serialize, Deserialize, Clone)]
struct PaymentResult {
    success: bool,
    transaction_id: Option<String>,
    error_message: Option<String>,
}

#[derive(Deserialize)]
struct ConfirmationRequest {
    id: String,
    result: PaymentResult,
}

// Mock function to initiate payment
async fn init_payment(req: &PaymentRequest, payment_id: &str) -> Result<String, RestateError> {
    println!(">>> Initiating external payment {}", payment_id);
    println!("  Confirm the payment via:");
    println!("  curl localhost:8080/PaymentsWithTimeout/confirm --json '{{\"id\": \"{}\", \"result\": {{\"success\": true, \"transactionId\": \"txn-123\"}}}}'", payment_id);
    Ok(format!("payRef-{}", Uuid::new_v4()))
}

// Mock function to cancel payment
async fn cancel_payment(pay_ref: &str) -> Result<(), RestateError> {
    println!(">>> Canceling external payment with ref {}", pay_ref);
    Ok(())
}

#[derive(Service)]
struct PaymentsWithTimeout;

impl PaymentsWithTimeout {
    #[restate_sdk::handler]
    async fn process(&self, ctx: &mut Context, req: PaymentRequest) -> Result<PaymentResult, RestateError> {
        let confirmation: Awakeable<PaymentResult> = ctx.awakeable();

        let pay_ref = {
            let payment_id = confirmation.id().clone();
            ctx.run("pay", async move {
                init_payment(&req, &payment_id).await
            }).await?
        };

        // Race between payment confirmation and timeout
        match timeout(Duration::from_secs(30), confirmation).await {
            Ok(result) => result,
            Err(_) => {
                // Timeout occurred - cancel the payment
                ctx.run("cancel-payment", async move {
                    cancel_payment(&pay_ref).await
                }).await?;
                
                Ok(PaymentResult {
                    success: false,
                    transaction_id: None,
                    error_message: Some("Payment timeout".to_string()),
                })
            }
        }
    }

    #[restate_sdk::handler]
    async fn confirm(&self, ctx: &mut Context, confirmation: ConfirmationRequest) -> Result<(), RestateError> {
        ctx.resolve_awakeable(&confirmation.id, confirmation.result)?;
        Ok(())
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let service = PaymentsWithTimeout;
    
    restate_sdk::serve(service)
        .bind("0.0.0.0:9080")
        .await?;
    
    Ok(())
}